src/Adaptor.js

```diff
@@ -197,6 +197,90 @@ export function upsert(table, fields) {
   };
 }
 
+/**
+ * Insert or update multiple records using ON DUPLICATE KEY
+ * @public
+ * @example
+ * upsertMany(
+ *   'users', // the DB table
+ *   [
+ *     { name: 'one', email: 'one@openfn.org' },
+ *     { name: 'two', email: 'two@openfn.org' },
+ *   ]
+ * )
+ * @constructor
+ * @param {string} table - The target table
+ * @param {array} data - An array of objects or a function that returns an array
+ * @returns {Operation}
+ */
+export function upsertMany(table, data) {
+  return function (state) {
+    /*
+    Builds and exexutes a query with syntax similar to the following 
+    (see MySQL official documentation for ON DUPLICATE KEY):
+     INSERT INTO users (name, email)
+    SELECT * FROM (
+      SELECT 'one' AS name, 'one@openfn.org' AS email UNION
+      SELECT 'two' AS name, 'two@openfn.org' AS email 
+    ) AS a
+    ON DUPLICATE KEY UPDATE name = a.name, email = a.email;
+    */
+    return new Promise(function (resolve, reject) {
+      const rows = expandReferences(data)(state); 
+
+      if (!rows || rows.length === 0) {
+        console.log('No records provided; skipping upsert.');
+        resolve(state);
+      }
+
+      const columns = Object.keys(rows[0]);
+      const columnList = columns.join(', '); 
+      
+      // 1. construct the INSERT statement
+      const insertStmt = "INSERT INTO ".concat(table, " (").concat(columnList, ")"); 
+      
+      // 2. construct the subquery
+      const subqueryAlias = 'a';
+      const selectValues = rows.map(function (r) {
+        let selectColumns = columns.map(function (c) {
+          return "".concat(mysql.escape(r[c]), " as ").concat(c);
+        }).join(', ');
+        return "SELECT ".concat(selectColumns);
+      });
+      const subqueryStmt = selectValues.join(' UNION ');
+      const selectStmt = "SELECT * FROM (".concat(subqueryStmt, ") as ").concat(subqueryAlias); 
+      
+      // 3. construct the ON DUPLICATE KEY statement
+      const updateStmt = columns.map(function (c) {
+        return "".concat(c, " = ").concat(subqueryAlias, ".").concat(c);
+      }).join(', ');
+      const onduplicateStmt = "ON DUPLICATE KEY UPDATE ".concat(updateStmt); 
+      
+      // combine 1, 2 & 3
+      const upsertString = "".concat(insertStmt, " ").concat(selectStmt, " ").concat(onduplicateStmt);
```

Comment:
I'm concerned why we aren't using the query builder here (and in other places), we're leaving people wide open to sql injection security issues.

The `squel` library used in this adaptor provides all the APIs required to build up this query (and ensure each input is sanitised:

```

var insertParams = squelMysql
    .insert({
      autoQuoteFieldNames: true,
    })
    .into(table)
    .setFields(valuesObj)
    .toParam()
    .onDupUpdate('field1 = field1 + 1')
    .onDupUpdate('field2', 'value2', {dummy: true})
    .onDupUpdate('field3', 'value3')
```

:point_up: this was lifted from the `upsert` function, with stuff I found in the `squel` library...
https://github.com/hiddentao/squel/blob/afa1cb52a61e89404e193bf4e36c2e79dd21c020/test/mysql.test.coffee#L58

Reply:
Using the `squel` library with multiple `onDupUpdate` statements is a good approach, I will give this a try.